File: examples/20260316-embed-go/kuu.go
// Package main demonstrates the embed+extract+exec pattern for kuu-cli integration.
//
// Architecture:
//
//	Host binary
//	  └── //go:embed kuu-cli  (platform-specific binary embedded at build time)
//	        └── extract to temp on first use
//	              └── exec: stdin JSON → stdout JSON
//
// The DX layer builds OptAST JSON, execs kuu-cli, and unmarshals the result.
// Users interact with a Go-native API — the subprocess is an implementation detail.
package main

import (
	"crypto/sha256"
	"embed"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
)

// binFS holds the embedded kuu-cli binaries.
// In a real build, this would contain platform-specific binaries:
//
//	bin/kuu-cli-linux-amd64
//	bin/kuu-cli-darwin-arm64
//	etc.
//
// For this PoC, we fall back to PATH lookup if no embedded binary is found.
//
//go:embed bin/*
var binFS embed.FS

var (
	resolvedPath string
	resolveOnce  sync.Once
	resolveErr   error
)

// resolveKuuCLI extracts the embedded kuu-cli binary to a cache directory,
// or falls back to PATH lookup. The binary is cached by content hash so
// extraction only happens once per version.
func resolveKuuCLI() (string, error) {
	resolveOnce.Do(func() {
		resolvedPath, resolveErr = doResolve()
	})
	return resolvedPath, resolveErr
}

func doResolve() (string, error) {
	// 1. Try embedded binary for this platform
	binName := fmt.Sprintf("bin/kuu-cli-%s-%s", runtime.GOOS, runtime.GOARCH)
	data, err := binFS.ReadFile(binName)
	if err == nil {
		return extractToCache(data)
	}

	// 2. Fall back to PATH
	path, err := exec.LookPath("kuu-cli")
	if err == nil {
		return path, nil
	}

	return "", fmt.Errorf("kuu-cli not found: no embedded binary for %s/%s and not in PATH", runtime.GOOS, runtime.GOARCH)
}

// extractToCache writes the binary to ~/.cache/kuu/kuu-cli-<hash>,
// skipping if already present.
func extractToCache(data []byte) (string, error) {
	hash := sha256.Sum256(data)
	hexHash := hex.EncodeToString(hash[:8]) // short hash is enough

	cacheDir, err := os.UserCacheDir()
	if err != nil {
		cacheDir = os.TempDir()
	}
	dir := filepath.Join(cacheDir, "kuu")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", fmt.Errorf("create cache dir: %w", err)
	}

	binPath := filepath.Join(dir, "kuu-cli-"+hexHash)
	if _, err := os.Stat(binPath); err == nil {
		return binPath, nil // already extracted
	}

	// Atomic write: write to temp, then rename
	tmp, err := os.CreateTemp(dir, "kuu-cli-*.tmp")
	if err != nil {
		return "", fmt.Errorf("create temp: %w", err)
	}
	defer os.Remove(tmp.Name())
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		return "", fmt.Errorf("write binary: %w", err)
	}
	tmp.Close()
	if err := os.Chmod(tmp.Name(), 0o755); err != nil {
		return "", fmt.Errorf("chmod: %w", err)
	}
	if err := os.Rename(tmp.Name(), binPath); err != nil {
		return "", fmt.Errorf("rename: %w", err)
	}
	return binPath, nil
}

// --- Schema types (same as WASM bridge JSON protocol v1) ---

type Schema struct {
	Version     int       `json:"version"`
	Description string    `json:"description,omitempty"`
	Opts        []OptDef  `json:"opts"`
	Args        []string  `json:"args"`
	RequireCmd  bool      `json:"require_cmd,omitempty"`
	Exclusive   [][]string `json:"exclusive,omitempty"`
	Required    []string  `json:"required,omitempty"`
}

type OptDef struct {
	Kind          string   `json:"kind"`
	Name          string   `json:"name"`
	Description   string   `json:"description,omitempty"`
	Shorts        string   `json:"shorts,omitempty"`
	Default       any      `json:"default,omitempty"`
	Global        bool     `json:"global,omitempty"`
	Hidden        bool     `json:"hidden,omitempty"`
	Aliases       []string `json:"aliases,omitempty"`
	Choices       []string `json:"choices,omitempty"`
	ImplicitValue any      `json:"implicit_value,omitempty"`
	Opts          []OptDef `json:"opts,omitempty"` // for command/serial
	RequireCmd    bool     `json:"require_cmd,omitempty"`
}

type ParseResult struct {
	OK            bool            `json:"ok"`
	Values        map[string]any  `json:"values,omitempty"`
	Command       *CommandResult  `json:"command,omitempty"`
	Error         string          `json:"error,omitempty"`
	Help          string          `json:"help,omitempty"`
	HelpRequested bool            `json:"help_requested,omitempty"`
}

type CommandResult struct {
	Name    string          `json:"name"`
	Values  map[string]any  `json:"values,omitempty"`
	Command *CommandResult  `json:"command,omitempty"`
}

// --- kuu-cli execution ---

// KuuParse sends a schema+args to kuu-cli and returns the parsed result.
// This is the core of the embed pattern — all the complexity is hidden here.
func KuuParse(schema *Schema) (*ParseResult, error) {
	kuuCLI, err := resolveKuuCLI()
	if err != nil {
		return nil, err
	}

	data, err := json.Marshal(schema)
	if err != nil {
		return nil, fmt.Errorf("marshal schema: %w", err)
	}

	cmd := exec.Command(kuuCLI, "parse")
	cmd.Stdin = strings.NewReader(string(data))
	cmd.Stderr = os.Stderr

	stdout, err := cmd.Output()
	if err != nil {
		// kuu-cli may exit non-zero for parse errors — check if we got JSON
		if exitErr, ok := err.(*exec.ExitError); ok && len(stdout) > 0 {
			_ = exitErr // we have stdout, try parsing it
		} else {
			return nil, fmt.Errorf("exec kuu-cli: %w", err)
		}
	}

	var result ParseResult
	if err := json.Unmarshal(stdout, &result); err != nil {
		return nil, fmt.Errorf("unmarshal result: %w (raw: %s)", err, stdout)
	}
	return &result, nil
}

// --- Convenience: self-re-exec pattern ---

// IsSelfKuuCLI checks if this process was invoked as an embedded kuu-cli.
// Used for the "multicall binary" pattern where the host program can also
// act as kuu-cli when invoked with a special env var.
//
// Usage in main():
//
//	if kuu.IsSelfKuuCLI() {
//	    kuu.RunAsKuuCLI()  // reads stdin, writes stdout, exits
//	}
//	// ... normal app logic
func IsSelfKuuCLI() bool {
	return os.Getenv("KUU_PARSE") == "1"
}

// RunSelfAsKuuCLI re-executes this binary with KUU_PARSE=1 to make it
// act as kuu-cli. This is an alternative to embedding a separate binary —
// the host program IS the kuu-cli.
//
// Requires: the host binary has kuu parsing logic compiled in (e.g., via
// MoonBit native → cgo, or a pure Go re-implementation of kuu core).
//
// For now, this is a placeholder showing the pattern.
func RunSelfAsKuuCLI() error {
	input, err := io.ReadAll(os.Stdin)
	if err != nil {
		return fmt.Errorf("read stdin: %w", err)
	}
	// In a real implementation, this would call the embedded kuu core.
	// For now, return an error indicating the pattern.
	_ = input
	return fmt.Errorf("RunSelfAsKuuCLI: not yet implemented (requires kuu core compiled into host binary)")
}

File: examples/20260316-embed-rust/src/kuu.rs
//! kuu-cli embed+extract+exec pattern for Rust.
//!
//! Embeds the kuu-cli binary via `include_bytes!`, extracts to a cache directory
//! on first use, and executes it as a subprocess for CLI argument parsing.
//!
//! # Architecture
//!
//! ```text
//! Host binary (Rust)
//!   └── include_bytes!("bin/kuu-cli")
//!         └── extract to ~/.cache/kuu/kuu-cli-{hash}
//!               └── exec: stdin JSON → stdout JSON
//! ```

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::io::Write;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::sync::OnceLock;

// --- Embedded binary ---

// Platform-specific binary selection at compile time.
// In production, use a build.rs to select the right binary.
#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
const KUU_CLI_BYTES: Option<&[u8]> = if cfg!(feature = "embed") {
    Some(include_bytes!("../bin/kuu-cli-linux-amd64"))
} else {
    None
};

#[cfg(not(all(target_os = "linux", target_arch = "x86_64")))]
const KUU_CLI_BYTES: Option<&[u8]> = None;

static KUU_CLI_PATH: OnceLock<Result<PathBuf, String>> = OnceLock::new();

/// Resolve the kuu-cli binary path. Extracts from embedded bytes or falls back to PATH.
fn resolve_kuu_cli() -> Result<PathBuf, String> {
    KUU_CLI_PATH
        .get_or_init(|| {
            // 1. Try embedded binary
            if let Some(bytes) = KUU_CLI_BYTES {
                return extract_to_cache(bytes);
            }

            // 2. Fall back to PATH
            which_kuu_cli()
        })
        .clone()
}

fn which_kuu_cli() -> Result<PathBuf, String> {
    let output = Command::new("which")
        .arg("kuu-cli")
        .output()
        .map_err(|e| format!("which kuu-cli: {e}"))?;
    if output.status.success() {
        let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
        Ok(PathBuf::from(path))
    } else {
        Err("kuu-cli not found in PATH".to_string())
    }
}

fn extract_to_cache(data: &[u8]) -> Result<PathBuf, String> {
    use std::os::unix::fs::PermissionsExt;

    let hash = {
        // Simple hash for cache key (first 16 bytes of content as hex)
        let preview: Vec<String> = data.iter().take(8).map(|b| format!("{b:02x}")).collect();
        preview.join("")
    };

    let cache_dir = dirs_cache().join("kuu");
    std::fs::create_dir_all(&cache_dir).map_err(|e| format!("create cache dir: {e}"))?;

    let bin_path = cache_dir.join(format!("kuu-cli-{hash}"));
    if bin_path.exists() {
        return Ok(bin_path);
    }

    // Atomic write
    let tmp_path = cache_dir.join(format!("kuu-cli-{hash}.tmp"));
    std::fs::write(&tmp_path, data).map_err(|e| format!("write binary: {e}"))?;
    std::fs::set_permissions(&tmp_path, std::fs::Permissions::from_mode(0o755))
        .map_err(|e| format!("chmod: {e}"))?;
    std::fs::rename(&tmp_path, &bin_path).map_err(|e| format!("rename: {e}"))?;

    Ok(bin_path)
}

fn dirs_cache() -> PathBuf {
    std::env::var("XDG_CACHE_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|_| {
            let home = std::env::var("HOME").unwrap_or_else(|_| "/tmp".to_string());
            PathBuf::from(home).join(".cache")
        })
}

// --- Schema types (kuu JSON protocol v1) ---

#[derive(Debug, Serialize)]
pub struct Schema {
    pub version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub opts: Vec<OptDef>,
    pub args: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub require_cmd: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exclusive: Option<Vec<Vec<String>>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required: Option<Vec<String>>,
}

#[derive(Debug, Default, Serialize)]
pub struct OptDef {
    pub kind: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub shorts: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub global: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hidden: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aliases: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub choices: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub implicit_value: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub opts: Option<Vec<OptDef>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub require_cmd: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ParseResult {
    pub ok: bool,
    pub values: Option<HashMap<String, serde_json::Value>>,
    pub command: Option<CommandResult>,
    pub error: Option<String>,
    pub help: Option<String>,
    pub help_requested: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CommandResult {
    pub name: String,
    pub values: Option<HashMap<String, serde_json::Value>>,
    pub command: Option<Box<CommandResult>>,
}

// --- kuu-cli execution ---

/// Send schema+args to kuu-cli, return parsed result.
pub fn kuu_parse(schema: &Schema) -> Result<ParseResult, String> {
    let kuu_cli = resolve_kuu_cli()?;
    let input = serde_json::to_string(schema).map_err(|e| format!("serialize schema: {e}"))?;

    let mut child = Command::new(&kuu_cli)
        .arg("parse")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .map_err(|e| format!("spawn kuu-cli: {e}"))?;

    // Write JSON to stdin
    if let Some(mut stdin) = child.stdin.take() {
        stdin
            .write_all(input.as_bytes())
            .map_err(|e| format!("write to kuu-cli stdin: {e}"))?;
    }

    let output = child
        .wait_with_output()
        .map_err(|e| format!("wait kuu-cli: {e}"))?;

    serde_json::from_slice(&output.stdout)
        .map_err(|e| format!("parse kuu-cli output: {e} (raw: {})", String::from_utf8_lossy(&output.stdout)))
}
